# **0.0 Imports**

In [20]:
import pandas as pd
import numpy as np
import seaborn as sns
import re
import sweetviz as sv
from sklearn import preprocessing as pp
from sklearn import decomposition as dd
import plotly.express as px
from sklearn.manifold import TSNE
from yellowbrick.cluster import KElbowVisualizer, SilhouetteVisualizer
from sklearn import cluster as c
from sklearn.metrics import silhouette_score
from sklearn import mixture as mx
import umap.umap_ as umap
from sklearn import manifold as mn
from sklearn import ensemble as en
from sklearn import metrics as m

## **0.2 Load Dataset**

In [2]:
data_raw = pd.read_csv('../data/raw/Ecommerce.csv', sep=',', decimal='.', encoding = 'ISO-8859-1' )
data_raw = data_raw.drop(columns = ['Unnamed: 8'])
df1 = data_raw.copy()

# **1.0 Descrição dos dados**

## 1.1 Rename Columns

In [3]:
cols_new = ['invoice_no', 'stock_code',
            'description', 'quantity', 'invoice_date', 
            'unit_price', 'customer_id', 'country']
df1.columns = cols_new

## 1.2 Data Dimensions

In [4]:
print('Number of rows: {}'.format(df1.shape[0]))
print('Number of columns: {}'.format(df1.shape[1]))

Number of rows: 541909
Number of columns: 8


## 1.3 Data types

In [5]:
df1.dtypes

invoice_no       object
stock_code       object
description      object
quantity          int64
invoice_date     object
unit_price      float64
customer_id     float64
country          object
dtype: object

## 1.4 Check NA

In [6]:
df1.isna().sum()

invoice_no           0
stock_code           0
description       1454
quantity             0
invoice_date         0
unit_price           0
customer_id     135080
country              0
dtype: int64

## 1.5 Replace NA

In [7]:
#substituir os NA
df_missing = df1.loc[df1['customer_id'].isna(), :] #pegou o df1 onde existem os NA e colocou em todas as colunas
df_not_missing = df1.loc[~df1['customer_id'].isna(), :] #pegou o df1 onde nao tem os na e colocou em todas as col

In [8]:
#rsubstituir os NA

#primeiro será na coluna customer_id [naoo tem 135082 registros]

df_backup = pd.DataFrame(df_missing['invoice_no'].drop_duplicates()) #removendo as duplicadas do invoice
df_backup['customer_id'] = np.arange(1900, 1900+len(df_backup),1) #pegou do df que tem NA e aplicou 1900 + o tamanho


#unindo o original com o df de referencia
df1 = pd.merge(df1, df_backup, on = 'invoice_no', how = 'left')

#coalesce

df1['customer_id'] = df1['customer_id_x'].combine_first(df1['customer_id_y']) #combinou. Se tem em uma e nao tem na outra ela cola

#apagar colunas extras

df1 = df1.drop(columns=['customer_id_x', 'customer_id_y'], axis = 1)

df1.head()


#depois do 19000 eu criei artificialmente

invoice_no stock_code                          description  quantity  \
0     536365     85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365      71053                  WHITE METAL LANTERN         6   
2     536365     84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365     84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365     84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

  invoice_date  unit_price         country  customer_id  
0    29-Nov-16        2.55  United Kingdom      17850.0  
1    29-Nov-16        3.39  United Kingdom      17850.0  
2    29-Nov-16        2.75  United Kingdom      17850.0  
3    29-Nov-16        3.39  United Kingdom      17850.0  
4    29-Nov-16        3.39  United Kingdom      17850.0

In [9]:
df1.isna().sum()

invoice_no         0
stock_code         0
description     1454
quantity           0
invoice_date       0
unit_price         0
country            0
customer_id        0
dtype: int64

## 1.6 Change types

In [10]:
df1['invoice_date'] =  pd.to_datetime(df1['invoice_date'], format='%d-%b-%y')
df1['customer_id'] = df1['customer_id'].astype(int)
df1.head()

invoice_no stock_code                          description  quantity  \
0     536365     85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365      71053                  WHITE METAL LANTERN         6   
2     536365     84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365     84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365     84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

  invoice_date  unit_price         country  customer_id  
0   2016-11-29        2.55  United Kingdom        17850  
1   2016-11-29        3.39  United Kingdom        17850  
2   2016-11-29        2.75  United Kingdom        17850  
3   2016-11-29        3.39  United Kingdom        17850  
4   2016-11-29        3.39  United Kingdom        17850

## 1.7 Descriptive Analysis

In [11]:
num_attributes = df1.select_dtypes(include = ['int64', 'float64'])
cat_attributes = df1.select_dtypes(exclude = ['int64', 'float64', 'datetime64[ns]'])

### 1.7.1 Numerical Atributes

In [12]:
# Central tendency - mean, median
#mean
ct1 = pd.DataFrame(num_attributes.apply(np.mean)).T
#median
ct2 = pd.DataFrame(num_attributes.apply(np.median)).T

#Dispersion - desvio, minimo, maximo, range, skew, kutosi

d1 = pd.DataFrame(num_attributes.apply(np.std)).T

d2 = pd.DataFrame(num_attributes.apply(np.min)).T

d3 = pd.DataFrame(num_attributes.apply(np.max)).T

d4 = pd.DataFrame(num_attributes.apply(lambda x: x.max() - x.min())).T

d5 = pd.DataFrame(num_attributes.apply(lambda x : x.skew())).T

d6 = pd.DataFrame(num_attributes.apply(lambda x: x.kurtosis())).T

#concat 
h = pd.concat([d2,d3,d4,ct1,ct2,d1,d5,d6]).T.reset_index()
h.columns = ['attributes', 'min', 'max', 'range', 'mean', 'median', 'std', 'skew', 'kurtosis'] 


h

attributes       min      max      range      mean  median         std  \
0    quantity -80995.00  80995.0  161990.00  9.552250    3.00  218.080957   
1  unit_price -11062.06  38970.0   50032.06  4.611114    2.08   96.759764   

         skew       kurtosis  
0   -0.264076  119769.160031  
1  186.506972   59005.719097

### 1.7.2 Categorical Attributes

In [21]:
df_letter_invoices = df1.loc[df1['invoice_no'].apply(lambda x: bool(re.search('[^0-9]+', x))), :]

In [22]:
print('total invoices with letter: {}'.format(len(df_letter_invoices)))
print('total quantity with negative value:  {}'.format(len(df_letter_invoices[df_letter_invoices['quantity'] < 0 ])))

total invoices with letter: 9291
total quantity with negative value:  9288


## **Stock Code**

In [23]:
df1.loc[df1['stock_code'].apply(lambda x: bool(re.search('^[a-zA-Z]+$', x))), 'stock_code'].unique()

array(['POST', 'D', 'DOT', 'M', 'S', 'AMAZONFEE', 'm', 'DCGSSBOY',
       'DCGSSGIRL', 'PADS', 'B', 'CRUK'], dtype=object)

In [24]:
df1['country'].value_counts(normalize = True).head()

United Kingdom    0.914320
Germany           0.017521
France            0.015790
EIRE              0.015124
Spain             0.004674
Name: country, dtype: float64

In [25]:
df1[['customer_id', 'country']].drop_duplicates().groupby('country').count().reset_index().sort_values('customer_id', ascending = False).head() 

country  customer_id
36  United Kingdom         7587
14         Germany           95
13          France           90
10            EIRE           44
31           Spain           31

# **2.0 Filter Variables**

In [26]:
df2 = df1.copy()

In [27]:
# === Numerical attributes ====
df2 = df2.loc[df2['unit_price'] >= 0.04, :]

# === Categorical attributes ====
df2 = df2[~df2['stock_code'].isin( ['POST', 'D', 'DOT', 'M', 'S', 'AMAZONFEE', 'm', 'DCGSSBOY', 'DCGSSGIRL', 'PADS', 'B', 'CRUK'] ) ]

# description
df2 = df2.drop( columns='description', axis=1 )

# map -  
df2 = df2[~df2['country'].isin( ['European Community', 'Unspecified' ] ) ]

# bad users
df2 = df2[~df2['customer_id'].isin( [16446] )]

# quantity
df2_returns = df2.loc[df1['quantity'] < 0, :]
df2_purchases = df2.loc[df1['quantity'] >= 0, :]

# **3.0 Feature Engineering**

In [28]:
df3 = df2.copy()

## **3.1 Feature Creation**

In [29]:
# data reference
df_ref = df3.drop( ['invoice_no', 'stock_code', 'quantity', 'invoice_date', 'unit_price', 'country'], axis=1 ).drop_duplicates( ignore_index=True )

### 3.1.1 Gross Revenue

In [30]:
 # Gross Revenue ( Faturamento ) quantity * price
df2_purchases.loc[:, 'gross_revenue'] = df2_purchases.loc[:, 'quantity'] * df2_purchases.loc[:, 'unit_price']

# Monetary
df_monetary = df2_purchases.loc[:, ['customer_id', 'gross_revenue']].groupby( 'customer_id' ).sum().reset_index()
df_ref = pd.merge( df_ref, df_monetary, on='customer_id', how='left' )
df_ref.isna().sum()

c:\Users\victo\Music\Projetos\insider_clustering\venv\lib\site-packages\pandas\core\indexing.py:1667: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[key] = value


customer_id       0
gross_revenue    91
dtype: int64

### **3.1.2 Recency**

In [31]:
 # Recency - Last day purchase
df_recency = df2_purchases.loc[:, ['customer_id', 'invoice_date']].groupby( 'customer_id' ).max().reset_index()
df_recency['recency_days'] = ( df2['invoice_date'].max() - df_recency['invoice_date'] ).dt.days
df_recency = df_recency[['customer_id', 'recency_days']].copy()
df_ref = pd.merge( df_ref, df_recency, on='customer_id', how='left' )
df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
dtype: int64

### **3.1.3 Quantity of purchased**

In [32]:
df_freq = (df2_purchases.loc[:, ['customer_id', 'invoice_no']].drop_duplicates()
                                                             .groupby( 'customer_id' )
                                                             .count()
                                                             .reset_index()
                                                             .rename( columns={'invoice_no': 'qtde_invoices'}) )
df_ref = pd.merge( df_ref, df_freq, on='customer_id', how='left' )
df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
qtde_invoices    91
dtype: int64


### **3.1.4 Quantity total of items purchased**m

In [33]:
# Numero de produtos
df_freq = (df2_purchases.loc[:, ['customer_id', 'quantity']].groupby( 'customer_id' ).sum()
                                                           .reset_index()
                                                           .rename( columns={'quantity': 'qtde_items'} ) )
df_ref = pd.merge( df_ref, df_freq, on='customer_id', how='left' )
df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
qtde_invoices    91
qtde_items       91
dtype: int64


### **3.1.5 Quantity of items purchased**

In [34]:
# Numero de produtos
df_freq = (df2_purchases.loc[:, ['customer_id', 'stock_code']].groupby( 'customer_id' ).count()
                                                           .reset_index()
                                                           .rename( columns={'stock_code': 'qtde_products'} ) )
df_ref = pd.merge( df_ref, df_freq, on='customer_id', how='left' )
df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
qtde_invoices    91
qtde_items       91
qtde_products    91
dtype: int64

### **3.1.6 Avg Ticket Value**

In [35]:
# Avg Ticket
df_avg_ticket = df2_purchases.loc[:, ['customer_id', 'gross_revenue']].groupby( 'customer_id' ).mean().reset_index().rename( columns={'gross_revenue':'avg_ticket'} )
df_ref = pd.merge( df_ref, df_avg_ticket, on='customer_id', how='left')
df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
qtde_invoices    91
qtde_items       91
qtde_products    91
avg_ticket       91
dtype: int64

### **3.1.6 Average Recency Days**

In [36]:
# Average recency days
df_aux = df2[['customer_id', 'invoice_date']].drop_duplicates().sort_values( ['customer_id', 'invoice_date'], ascending=['False', 'False'] )
df_aux['next_customer_id'] = df_aux['customer_id'].shift() # next customer
df_aux['previous_date'] = df_aux['invoice_date'].shift() # next invoince date

df_aux['avg_recency_days'] = df_aux.apply( lambda x: ( x['invoice_date'] - x['previous_date'] ).days if x['customer_id'] == x['next_customer_id'] else np.nan, axis=1 )

df_aux = df_aux.drop( ['invoice_date', 'next_customer_id', 'previous_date'], axis=1 ).dropna()

# average recency 
df_avg_recency_days = df_aux.groupby( 'customer_id' ).mean().reset_index()

# merge
df_ref = pd.merge( df_ref, df_avg_recency_days, on='customer_id', how='left' )
df_ref.isna().sum()

customer_id            0
gross_revenue         91
recency_days          91
qtde_invoices         91
qtde_items            91
qtde_products         91
avg_ticket            91
avg_recency_days    2816
dtype: int64

### **3.1.7 Frequency Purchase**

In [37]:
df_aux = ( df2_purchases[['customer_id', 'invoice_no', 'invoice_date']].drop_duplicates()
                                                             .groupby( 'customer_id')
                                                             .agg( max_ = ( 'invoice_date', 'max' ), 
                                                                   min_ = ( 'invoice_date', 'min' ),
                                                                   days_= ( 'invoice_date', lambda x: ( ( x.max() - x.min() ).days ) + 1 ),
                                                                   buy_ = ( 'invoice_no', 'count' ) ) ).reset_index()
# Frequency
df_aux['frequency'] = df_aux[['buy_', 'days_']].apply( lambda x: x['buy_'] / x['days_'] if  x['days_'] != 0 else 0, axis=1 )

# Merge
df_ref = pd.merge( df_ref, df_aux[['customer_id', 'frequency']], on='customer_id', how='left' )

df_ref.isna().sum()

customer_id            0
gross_revenue         91
recency_days          91
qtde_invoices         91
qtde_items            91
qtde_products         91
avg_ticket            91
avg_recency_days    2816
frequency             91
dtype: int64

### **3.1.8 Number of Returns**

In [38]:
 # Number of Returns
df_returns = df2_returns[['customer_id', 'quantity']].groupby( 'customer_id' ).sum().reset_index().rename( columns={'quantity':'qtde_returns'} )
df_returns['qtde_returns'] = df_returns['qtde_returns'] * -1

df_ref = pd.merge( df_ref, df_returns, how='left', on='customer_id' )
df_ref.loc[df_ref['qtde_returns'].isna(), 'qtde_returns'] = 0

df_ref.isna().sum()

customer_id            0
gross_revenue         91
recency_days          91
qtde_invoices         91
qtde_items            91
qtde_products         91
avg_ticket            91
avg_recency_days    2816
frequency             91
qtde_returns           0
dtype: int64

# **4.0 Exploratory Data Analysis (EDA)**

In [39]:
df4 = df_ref.dropna()
df_ref.shape

(5786, 10)

## **4.3 Space Study**

In [40]:
# original dataset
#df43 = df4.drop( columns=['customer_id'], axis=1 ).copy()

# selected dataset
cols_selected = ['customer_id', 'gross_revenue', 'recency_days', 'qtde_products', 'frequency', 'qtde_returns']
df43 = df4[ cols_selected ].copy() 

In [41]:
mm = pp.MinMaxScaler()
# 

df43['gross_revenue'] = mm.fit_transform( df43[['gross_revenue']] )
df43['recency_days']           = mm.fit_transform( df43[['recency_days']] )
#df43['qtde_invoices']          = mm.fit_transform( df43[['qtde_invoices']])
#df43['qtde_items']             = mm.fit_transform( df43[['qtde_items']])
df43['qtde_products']          = mm.fit_transform( df43[['qtde_products']])
#df43['avg_ticket']             = mm.fit_transform( df43[['avg_ticket']])
#df43['avg_recency_days']       = mm.fit_transform( df43[['avg_recency_days']])
df43['frequency']              = mm.fit_transform( df43[['frequency']])
df43['qtde_returns']           = mm.fit_transform( df43[['qtde_returns']])
#df43['avg_basket_size']        = mm.fit_transform( df43[['avg_basket_size']])
#df43['avg_unique_basket_size'] = mm.fit_transform( df43[['avg_unique_basket_size']] )


X = df43.copy()

In [42]:
X.shape

(2968, 6)

### **4.4.1 Tree-Based Embedding**

In [43]:
# training dataset
X = df43.drop( columns=['customer_id', 'gross_revenue'], axis=1 )
y = df43['gross_revenue']

# model definition
rf_model = en.RandomForestRegressor( n_estimators=100, random_state=42 )

# model training
rf_model.fit( X, y )

# Leaf 
df_leaf = pd.DataFrame( rf_model.apply( X ) )
# dataframe Leaf

In [44]:
# Reduzer dimensionality
reducer = umap.UMAP( random_state=42 )
embedding = reducer.fit_transform( df_leaf )

# embedding
df_tree = pd.DataFrame()
df_tree['embedding_x'] = embedding[:, 0]
df_tree['embedding_y'] = embedding[:, 1]

# **5.0 Data Preparation**

In [45]:
df5 = df_tree.copy()

# **7.0 Hyperparameter Fine-Tunning**

In [46]:
X = df_tree.copy()

In [47]:
X.head()

embedding_x  embedding_y
0    -0.408562    -8.277472
1    -7.735217    -2.287607
2     0.551945    -8.545471
3     8.924150     6.583935
4    14.244323     2.876050

# **8.0 Model Training**

In [48]:
# model definition
k = 8

gmm_model = mx.GaussianMixture( n_components=k, n_init=300, random_state=42)

# model training
gmm_model.fit( X )

# model predict
labels = gmm_model.predict( X )

In [49]:
print('SS value:{}'.format(m.silhouette_score(X, labels, metric = 'euclidean')))

SS value:0.5836340188980103


# **9.0 Cluster Analysis**

In [50]:
df92 = df4[ cols_selected ].copy()
df92['cluster'] = labels

#change dtypes
df92['recency_days'] = df92['recency_days'].astype(int)
df92['qtde_products'] = df92['qtde_products'].astype(int)
df92['qtde_returns'] = df92['qtde_returns'].astype(int)



In [51]:
# Number of customer
df_cluster = df92[['customer_id', 'cluster']].groupby( 'cluster' ).count().reset_index()
df_cluster['perc_customer'] = 100*( df_cluster['customer_id'] / df_cluster['customer_id'].sum() )

# Avg Gross revenue
df_avg_gross_revenue = df92[['gross_revenue', 'cluster']].groupby( 'cluster' ).mean().reset_index()
df_cluster = pd.merge( df_cluster, df_avg_gross_revenue, how='inner', on='cluster' )

# Avg recency days
df_avg_recency_days = df92[['recency_days', 'cluster']].groupby( 'cluster' ).mean().reset_index()
df_cluster = pd.merge( df_cluster, df_avg_recency_days, how='inner', on='cluster' )

# Avg invoice_no
df_qtde_products = df92[['qtde_products', 'cluster']].groupby( 'cluster' ).mean().reset_index()
df_cluster = pd.merge( df_cluster, df_qtde_products, how='inner', on='cluster' )

# Frequency
df_frequency = df92[['frequency', 'cluster']].groupby( 'cluster' ).mean().reset_index()
df_cluster = pd.merge( df_cluster, df_frequency, how='inner', on='cluster' )

# Returns
df_qtde_returns = df92[['qtde_returns', 'cluster']].groupby( 'cluster' ).mean().reset_index()
df_cluster = pd.merge( df_cluster, df_qtde_returns, how='inner', on='cluster' )

df_cluster.sort_values('gross_revenue', ascending = False)

cluster  customer_id  perc_customer  gross_revenue  recency_days  \
1        1          468      15.768194    8836.136389     21.438034   
0        0          130       4.380054    4689.930154     47.407692   
7        7          145       4.885445    3164.253379     34.427586   
4        4          760      25.606469    2024.409513     49.556579   
6        6          376      12.668464    1240.908644     62.840426   
3        3          322      10.849057     971.025435     75.329193   
2        2          200       6.738544     647.933500     47.280000   
5        5          567      19.103774     605.034215    131.708995   

   qtde_products  frequency  qtde_returns  
1     424.134615   0.094571    149.271368  
0     103.900000   0.057342     20.753846  
7     173.875862   0.060415     23.482759  
4     107.551316   0.046418     15.459211  
6      53.904255   0.046065     15.555851  
3      37.462733   0.074586      6.571429  
2      11.510000   0.025991      0.845000  
5      19.010582   0.344916     13.587302

In [148]:
#1 - Insiders
#0
#7
#4
#6
#3
#2
#5
#
#


## Cluster 01 (Candidato à insiders)

- Número de customers: 468 (16% do customers )
- Faturamento médio: 8836
- Recência média: 21 dias
- Média de Produtos comprados: 424 produtos
- Frequência de Produtos comprados: 0.09 produtos/dia
- Receita em média: $8836.13,00 dólares

## Cluster 02 

- Número de customers: 31 (0.71 do customers )
- Recência em média: 14 dias
- Compras em média: 53 compras
- Receita em média: $40.543,52 dólares

## Cluster 03

- Número de customers: 4.335 (99% do customers )
- Recência em média: 92 dias
- Compras em média: 5 compras
- Receita em média: $1.372,57 dólares


# **11.0 Deploy to Production**

In [52]:
df92.dtypes

customer_id        int32
gross_revenue    float64
recency_days       int32
qtde_products      int32
frequency        float64
qtde_returns       int32
cluster            int64
dtype: object

## **11.1 Inserting into SQLite**

In [53]:
import sqlite3
from sqlalchemy import create_engine

In [55]:
#create table
query_create_table_insiders = """
   CREATE TABLE insiders(
       customer_id      INTEGER,
       gross_revenue    REAL,
       recency_days     INTEGER,
       qtde_products    INTEGER,
       frequency        REAL,
       qtde_returns     INTEGER,
       cluster          INTEGER
   )
"""
conn = sqlite3.connect('insiders_db.sqlite')
conn.execute(query_create_table_insiders)
conn.commit()
conn.close()
#insert data
conn = create_engine('sqlite:///C:\\Users\\victo\\Music\\Projetos\\insider_clustering\\notebooks\\insiders_db.sqlite')
df92.to_sql('insiders', con = conn, if_exists='append', index=False)
#select data



In [56]:
query = """
    select * from insiders
"""
df = pd.read_sql_query(query, conn)


In [58]:
df.head()

customer_id  gross_revenue  recency_days  qtde_products  frequency  \
0        17850        5391.21           372            297  17.000000   
1        13047        3232.59            56            171   0.028302   
2        12583        6705.38             2            232   0.040323   
3        13748         948.25            95             28   0.017921   
4        15100         876.00           333              3   0.073171   

   qtde_returns  cluster  
0            40        1  
1            35        7  
2            50        1  
3             0        5  
4            22        5